### PR TITLE
Issue #12600: Modify javadocstyle check to work with inline return

### PR DESCRIPTION
**Pull request by 0xbakry**
Solves https://github.com/checkstyle/checkstyle/issues/12600

Since inline `@return` tag is the only inline tag that automatically inserts a period after the text, we can modify check method to determine if the first sentence was `{@return something}`

CLI:
```
$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="JavadocStyle"/>
    </module>
</module>

$ cat Test.java
class Test {
    /**
     * {@return a string}
     */
    String foo() {
        return "Hello, world";
    }
}

$ java -jar checkstyle-10.10.1-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```

Diff Regression projects: https://gist.githubusercontent.com/0xbakry/7853843aa3d69aee301d7a8d62d08ebb/raw/2e52c13c579d1461e515394bbcbc752ee9c519cc/my_projects
Diff Regression config: https://gist.githubusercontent.com/0xbakry/a1fcb13eef42f29de8ff500892f34166/raw/6c59fb003381d5634486a2b908d1f0463515fb58/config_check.xml

Clean report:
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/7324fc0_2023132220/reports/diff/index.html

### CODE REVIEW

**Comment by 0xbakry** [comment]
GitHub, generate report

**Comment by 0xbakry** [comment]
@romani @nrmancuso Please give me your feedbacks.

For the failing check, I can't fully catch/understand `ci/semaphoreci/pr: Checkstyle CI pipeline on Semaphore`, do I need to re-arrange the order of methods?

**Comment by romani** [comment]
I restarted semaphore, I fixed distance project problems.

**Comment by 0xbakry** [comment]
@romani Thanks! I'm waiting for your feedback.

**Comment by romani** [comment]
@0xbakry  @nrmancuso , please read https://github.com/checkstyle/checkstyle/issues/12600#issuecomment-1529114076

We will merge this PR, but it might be much better to completely remove such validation from this Check to not confuse user, there are much better Check to validate it. Sorry that I didn't catch it on moment of issue approval.

**Review by romani — APPROVED** [review]
As there is no much regression, ok to merge

**Comment by 0xbakry** [comment]
@nrmancuso Ping

File: config/pitest-suppressions/pitest-javadoc-suppressions.xml
Comment (nrmancuso) [inline config/pitest-suppressions/pitest-javadoc-suppressions.xml:535]: @romani to be clear, you are good to extend list of surviving mutations?

Comment (romani) [reply config/pitest-suppressions/pitest-javadoc-suppressions.xml:535]: No extension is allowed for technical dept.
Especially for this questionable update, as we have better Check to validate it.

Comment (romani) [reply config/pitest-suppressions/pitest-javadoc-suppressions.xml:535]: @0xbakry, please update tests to cover this mutation survival.

Comment (0xbakry) [reply config/pitest-suppressions/pitest-javadoc-suppressions.xml:535]: Sorry for delay, I was a little bit busy with my final exams, I'll try to update it soon.

Comment (romani) [reply config/pitest-suppressions/pitest-javadoc-suppressions.xml:535]: @0xbakry, ping, do we have chance to finish this PR

Comment (0xbakry) [reply config/pitest-suppressions/pitest-javadoc-suppressions.xml:535]: I will close it for now, till I finish my GP and I'll be back.

**Comment by romani** [comment]
CI is red , please fix it.

**Comment by 0xbakry** [comment]
> CI is red , please fix it.

Yes, I'm working on it, there are many changes in the project so I just need some time to work correctly again, Thanks!

**Comment by romani** [comment]
You are close to green CI.
Still 2 items to cover by tests https://github.com/checkstyle/checkstyle/actions/runs/7556390072/job/20573295075?pr=13037#step:7:23

**Comment by 0xbakry** [comment]
@romani I'm not sure if this test case will cover them correctly.
```
   /**
    * Sentence with whitespace at the end. 
    */
 ```
 On the other side, this will not build successfully due to `Trailing whitespace is not allowed [noTrailingWhitespace]`

**Comment by romani** [comment]
>Trailing whitespace is not allowed [noTrailingWhitespace]

We usually do not run checkstyle on Inputs, as they are weird by design. The more weird the better.

**Comment by nrmancuso** [comment]
> You are close to green CI.

@romani we are missing almost half of our CI workflows in this PR.

**Comment by 0xbakry** [comment]
> > You are close to green CI.
> 
> @romani we are missing almost half of our CI workflows in this PR.
Sorry, i didn't get it، how "half"?

**Comment by nrmancuso** [comment]
@0xbakry it looks like at least CircleCI is not running on this PR, they have some sanctions against IPs in certain countries

**Comment by romani** [comment]
@0xbakry, please make other CI green. At the end maintainers can copy your code to other PR to check circleci

**Comment by 0xbakry** [comment]
@romani @nrmancuso Sorry for the delay. This red CI is not happening cause of my work, right?

**Comment by rnveach** [comment]
@romani 2 CI failures do not look related to the code in this PR. First one doesn't show a failure, other is our limit is exceeded.

**Comment by romani** [comment]
Yes, non of them related, just known CI issues

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/InputJavadocStyleDefaultSettingsFive.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/InputJavadocStyleDefaultSettingsFive.java:21]: >// ok, foo one

Please remove it. Or add javadoc that was in issue. It is good practice to show new behavior on what is reported in issue.

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/InputJavadocStyleDefaultSettingsFive.java
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/InputJavadocStyleDefaultSettingsFive.java:21]: All comment text that try to explain violation in javadoc , better to place outside of javadoc

**Comment by romani** [comment]
GitHub, generate report

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/InputJavadocStyleDefaultSettingsFive.java
Comment (0xbakry) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/InputJavadocStyleDefaultSettingsFive.java:21]:
```
   /**
    * {@return a string}
    */
    public String foo() // ok, foo one
    {
        return "Hello, world";
    }
```
is this fine?

**Review by romani — CHANGES_REQUESTED** [review]
In config for diff report there is no module JavadocStyle. Please remove unrelated modules and use single Check. And please extend list of projects, put jdk in list of projects.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/InputJavadocStyleDefaultSettingsFive.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/InputJavadocStyleDefaultSettingsFive.java:23]: Please remove such ok comments, they don't explain anything.
Remove all others that you introduce in this PR

**Review by romani — CHANGES_REQUESTED** [review]
Items

**Comment by 0xbakry** [comment]
> In config for diff report there is no module JavadocStyle. Please remove unrelated modules and use single Check. And please extend list of projects, put jdk in list of projects.

I updated them, please take a look at them and tell me if this is correct

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/InputJavadocStyleDefaultSettingsFive.java
Comment (0xbakry) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/InputJavadocStyleDefaultSettingsFive.java]: Done.

**Comment by 0xbakry** [comment]
GitHub, generate report

**Comment by romani** [comment]
ok, code is ok, lets wait for diff report that will prove no regression.

**Comment by romani** [comment]
Please add to your input file cases:
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/658bb6b_2024172041/reports/diff/openjdk17/index.html#A2

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/658bb6b_2024172041/reports/diff/openjdk17/index.html#A5

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/658bb6b_2024172041/reports/diff/openjdk17/index.html#A49

After update no need to regenerate report. Report is good , I don't see problems, great job!

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java:352]: Please make this a compiled pattern also.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java:219]: This can be static.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java:357]: This can be the result of one immutable stream:
```suggestion
        final boolean isInLineReturnTag = Arrays.stream(WORD_DELIMITER.split(commentText))
                .findFirst()
                .map(inlineReturnTagPattern::matcher)
                .filter(Matcher::find)
                .isPresent();
```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java:361]:
```suggestion
            && !isInLineReturnTag
```

**Review by nrmancuso — CHANGES_REQUESTED** [review]
We need to extend all documentation examples to include inline return tags, it may not be obvious to users why we don't check this.

Items:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (0xbakry) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java:361]: Sorry but I didn't get it, what is the point of such updates? why use FP in this simple logic?

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java:361]: code should be a English text, especially in long IF expressions. Non of like to read code for long to understand it, we try to come to model where we(any other contributr) read code quickly by junior engineer.

Comment (0xbakry) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java:361]: yeah that's exactly why I think using FP will be more complex. it's ok for me to apply these changes, i'm just saying the original code is straight forward than the updated one. so please take another look at both and confirm if we really want to update the code.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java:361]: I like proposal of @nrmancuso .

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java:361]: >Sorry but I didn't get it, what is the point of such updates? why use FP in this simple logic?

Good question :) My suggestion has nothing to do with using or not using functional programming. Here is why I have suggested such changes:

1. Maintainability. Our intention is immediately obvious to the next person to read this code. Imagine you come back to look at this code in a few months:

` final String[] sentences = commentText.split("\\.(?=\\s|$)");`

I see we want to split comment content by punctuation and whitespace into sentences

`String firstSentence = "";`

We are declaring a blank string for some reason...

```
        if (sentences.length > 0) {
            firstSentence = sentences[0];
        }
```

Ok, we want to get the first sentence and possibly reassign the `firstSentence` variable.

```
     if (!commentText.isEmpty()
            && !endOfSentenceFormat.matcher(commentText).find()
            && !inlineReturnTagPattern.matcher(firstSentence).find()
```

After re-reading these conditions a few times, I see that we want to use the inline return tag pattern matcher on the first sentence from the comment, and negate this. Ah - we want to make sure that the first sentence in this comment does not have an inline return tag!

Versus:

`final boolean hasInLineReturnTag...`

I don't need to read the rest of the stream operations. We want to know if the first sentence has an inline return tag. 

We visit the condition below:

```
        if (!commentText.isEmpty()
            && !endOfSentenceFormat.matcher(commentText).find()
            && !hasInLineReturnTag
```
We want to make sure this first sentence does NOT have an inline return tag.

2. Immutability. We should always prefer immutable data whenever possible.
3. Concision. We don't care about or need the result of any of the intermediate operations, just the final outcome.
4. From a high level, we want to know if some element of a collection matches our criteria. Java has an API for this; the Stream API. We should use it.

Comment (0xbakry) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java:361]: @nrmancuso great, thanks for that!
now, do we need to extend the documentation examples in this PR? or should we open another issue for it?

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java:361]: We should extend documentation in this PR.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (0xbakry) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (0xbakry) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (0xbakry) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (0xbakry) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java]: Done.
While updating the documentation I found this wrong [Example](https://checkstyle.org/checks/javadoc/javadocstyle.html#:~:text=private%20void%20methodWithValidCommentStyle()%20%7B%7D%0A%0A%20%20%20%20/**-,*%20Some%20description%20here%20//%20violation%2C%20the%20sentence%20must%20end%20with%20a%20proper%20punctuation,-*/%0A%20%20%20%20private%20void%20methodWithInvalidCommentStyle), this doesn't produce a violation and I don't know if it was intended or not, anyway I updated it in the documentation for now.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java]: Bad naming on my part, please change this to `SENTENCE_SEPARATOR`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java]: Please restore what was there, and let's open a new issue for this. We can mention that documentation is misleading and that examples need to be updated in the new issue.

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt
Comment (nrmancuso) [inline src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt:19]: This example doesn't make sense; we would never have a return tag on a void method. Please update this method to return a `String`

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Last item:

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt
Comment (romani) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt:19]: and make it somthing like:
```
// ok below, @return tag automatically inserts a period after the text
/**
     * {@return something} 
     */
```

to keep javadoc comment "realistic" and all hints text outside to never affect validatin.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (0xbakry) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java]: Done at https://github.com/checkstyle/checkstyle/issues/14510

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (0xbakry) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java]: Done.

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt
Comment (nrmancuso) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt:19]: @0xbakry please also update the first item under “ The following checks are performed:” in the check description

Comment (0xbakry) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt]: Totally agree!! I have missed this, sorry. 
Done.

Comment (0xbakry) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt]: > @0xbakry please also update the first item under “ The following checks are performed:” in the check description

Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java:55]:
```suggestion
 * Note that this check is not applied to inline <code>{@return ...}</code> tags,
 * because the Javadoc tools automatically appends a period to the end of the tag 
 * content.
```

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Let's be very clear about check behavior here, `@return` appears to mention the typical return tag, not the inline tag:

**Comment by 0xbakry** [comment]
GitHub, generate website

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java]: >I think we're oversimplifying things.

That's ok; not everyone is a java expert. I certainly didn't know about this nuance before this PR.

>people who will read the documentation are developers by the way

Yes, I am aware of this fact. 

>the below examples will show them how it exactly works

It is good to go into a bit more detail in the description like we have here, and mentioning some important detail of this check's behavior a couple of different ways is good; everyone fails to notice things in examples from time to time.

**Review by nrmancuso — APPROVED** [review]
i am good with the changes here, but we need a maintainer to open a PR on this branch to make sure that all the missing CI is passing.

**Comment by nrmancuso** [comment]
@romani please complete review

**Comment by nrmancuso** [comment]
> i am good with the changes here, but we need a maintainer to open a PR on this branch to make sure that all the missing CI is passing.

PR is opened at https://github.com/checkstyle/checkstyle/pull/14528

**Comment by nrmancuso** [comment]
@0xbakry please address inspections failures at https://app.circleci.com/pipelines/github/checkstyle/checkstyle/24047/workflows/c4bc7400-3c5b-44dc-8ac2-bf30335f74a7/jobs/524911/artifacts

**Comment by nrmancuso** [comment]
@romani please continue review

**Comment by nrmancuso** [comment]
@0xbakry we have several CI failures that must be addressed at https://github.com/checkstyle/checkstyle/pull/14536, please do.

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt
Comment (romani) [inline src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt:20]:
```
{@return {@code true} if this object has been initialized, {@code false} otherwise}
```

And 
`private boolean isInitialized(){ return true; }`

Goal is to show user good example of usage, as nuance with this tag was surprise even for me. I didn't not know how to use it in this way.

**Review by romani — CHANGES_REQUESTED** [review]
Absolutely last improvement:

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt
Comment (0xbakry) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt:20]: excuse me if I didn't get your point, do you want to add this example to the documentation?

**Comment by 0xbakry** [comment]
@romani Ping

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt
Comment (0xbakry) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt]: @romani all changes were done in the last commit, please can you finish the review?

Comment (0xbakry) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/Example1.txt]: @romani Ping

**Comment by romani** [comment]
GitHub, generate website

File: src/xdocs/checks/javadoc/javadocstyle.xml
Comment (romani) [inline src/xdocs/checks/javadoc/javadocstyle.xml:211]: Please make it single line method to use less vertical space.
Please add this chunk to all other examples

**Review by romani — CHANGES_REQUESTED** [review]
Last

**Review by romani — APPROVED** [review]
Ok to merge.

Examples as not migrated yet

**Pull request merged**
